README.md
# 'game'
### By Yakasov
I made this mostly during quarantine. Just a fun small project.


## Version descriptions
Below is all versions committed before changing to a different commit naming scheme. The change count is increased by 1 for every type, command, item etc. added in that update

eg 3 commands added represents 3 changes


Version Number | Description | Changes
---------------|-------------|---------
1.1 | Enemy projectile system + run.bat addition | 11
1.2.1 | Enemy, boss randomisation + commands | 8
1.3 | GUI elements + game logic tweaks | 14
1.3.1 | Window title | 2
1.3.2 | Minor code tweaks | 1
1.4 | Item bug fixes | 14
1.4.1 | QoL gameplay speed changes | 8
1.4.2 | Minor code tweaks | 1
1.4.3 | Bug fixes | 2
1.4.4 | Small polish tweaks | 6
1.5 | Balance changes + polish | 18
1.5.1 | Changelog fix | 2
1.6 | Code improvements | 6
1.6.1 | Small polish tweaks | 2
1.7 | New items | 7


changelog.txt
? Add End Boss
? Add events
? Add special rooms, NPCs? Messages?
? Sound? Graphics?

FOR VERSION TITLES FOUND ON GITHUB PRIOR TO 1.7.1, PLEASE REFER TO README

1.8.3
~ Decreased PROJECTILE VELOCITY magnitude change to 1.5x from 2x
~ Increased PROJECTILE RATE magnitude change to 8x from 5x
~ Decreased all item chances by 20%
~ Enemies have an even chance of becoming any enemy type, rather than 33% becoming shooter type specifically
~ Increased enemy base velocity by 16%
~ Enemies with type 'base' move 1.2x faster than normal
~ Bosses will fire more random shots
~ Increased enemy death particle duration
~ More particles are produced upon enemy death
* Added error handling for rare enemy removal crash

1.8.2
* Fixed projectiles being incorrectly removed when leaving the screen
* Fixed player projectiles sometimes registering damage multiple times

1.8.1
+ Added map feature, viewable using TAB
+ Bosses will spawn with enemies
+ Enemy sizes will now vary (but will all have the same size ratio)
~ Increased GOLD ALL chance magnitude from 1 to 2
~ Increased PLAYER_HEALTH item bonuses by 50%
~ Increased player projectile speed by 33%
~ Enemies fire 33% faster
~ Boss movement speed range increased (from 1 -> 4 to 0.66 -> 4.33)
  Boss movement speed increase over time increased by 2x
~ Bosses will vary between firing accurately and firing randomly in a cone towards you
~ Boss fire rate increased (now 5ms between shots)
~ Boss health maximum decreased
~ Map generator changes:
  Any given tile now has a 50% chance of becoming a room, rather than 70%
  Map patching has been made less likely (previously if a tile had 2 neighbours,
  it would also become a room. This now happens 33% of the time instead)
~ GREY colour made lighter
* Boss room chance fixed as was slightly too high
! Maps can generate with inaccessible rooms, viewable on the map
! Enemies can phase through bosses - they can be killed normally when they move outside

1.8
+ Added new enemy kill animation
+ Added acquired items display
+ Added end screen score display
+ Added new items:
  WHITE PLAYER_VELOCITY: +1 velocity, 2% chance
  WHITE PLAYER_HEALTH: +10 health, 2% chance
  WHITE PROJECTILE_VELOCITY: +3 projectile velocity, 2% chance
  WHITE PROJECTILE_RATE: -25ms fire time, 2% chance
  GREY PROJECTILE_RATE: -10ms fire time, 15% chance
  GREY PROJECTILE_VELOCITY: +1.5 projectile velocity, 15% chance
~ Player model size increased by 33%
~ Player speeds increased
~ Player enters a screen closer to the border
~ Enemies will now only give health 33% of the time
~ Enemy projectiles size increased by 33%
~ Enemies move 25% faster
~ Enemy fire rate increases faster
~ Enemies spawn closer to the middle of the screen
~ Bosses heal less health
~ Bosses deal less damage from collisions
~ Bosses fire 20% faster
~ Boss projectiles are 25% more accurate
~ Map generator changes:
  Any given tile now has a 70% chance of becoming a room, rather than 80%
  Any given tile has a 20% chance of becoming a boss room
~ Inverted controls:
  WASD is now used for movement
  Arrow Keys are now used for firing
~ items.dat syntax simplified
~ Increased LIME PROJECTILE_RATE item chance to 7% (from 5%)
~ Window size changed from 500x500 to 750x750
* Fixed bossProjFireRate being used incorrectly causing too much boss projectile firing
* Fixed enemy projectiles persisting across screens

================================================================================

1.7.1
+ Health is now also shown in text form on screen
~ AQUA PLAYER_VELOCITY item magnitude decreased to 1 (from 1.5)
- Removed some unnecessary event calls
* Projectiles now cause invulnerability to trigger

1.7
+ Added new items:
  GOLD ALL = 1% chance, affects PLAYER_VELOCITY PLAYER_HEALTH PROJECTILE_VELOCITY PROJECTILE_RATE by magnitude 1
  (+1 velocity, +10 health, +2 projectile velocity, -5% fire time)
  LIME RATE = 5% chance, affects PROJECTILE_RATE by magnitude 3 (-15% fire time)
~ Increased AQUA PLAYER_VELOCITY item chance to 8% (from 5%)
~ Increased MAGENTA PROJECTILE_VELOCITY item chance to 15% (from 8%)
~ Changed colours.dat to colours.py
  Colours can now be called directly using their name
  eg RED instead of COLOURS['RED']
~ Items affecting player health are now 10x their magnitude
~ Added all existing tuple colours in code to colours.py

================================================================================

1.6.1
+ You can no longer fire once a room has been cleared
* Fixed debug info printing cleared velocity rather than normal velocity for the player

1.6
+ Added a new class, render()
~ All functions involved in drawing objects have been moved to render()
~ Some code split into smaller functions
* Fixed invulnerability not stopping projectile damage
* Fixed incorrect change note:
  '~ Projectile velocity decreased to 6 (320u/s) (from 8 (480u/s))', Version 1.5
  New velocity is 360u/s, not 320u/s

================================================================================

1.5.1
+ Added ignore rule to .pylintrc
* Fixed a line being missing in the changelog

1.5
+ If an item is on a previous cleared screen, the border to it will appear gold instead of the default grey
+ Added a new command:
  genmap [int]: generates maps in console [int] times
+ Pressing 'x' will print debug info at any point
  Debug info still prints when entering a screen
+ Bosses now fire projectiles with a degree of randomness to make them harder to dodge
+ Added intro text to console on game startup
  Debug info is no longer printed on screen 11
+ Added proper file process termination when files have finished being used.
~ Player projectile fire rate decreased to 6/s (from 10/s)
~ Player velocity increased to 4.5 (270u/s) (from 4 (240u/s))
~ Projectile velocity decreased to 6 (320u/s) (from 8 (480u/s))
~ AQUA PLAYER_VELOCITY item magnitude increased to 1.5 (from 1) - this equates to a 90u/s bonus
~ MAGENTA PROJECTILE_VELOCITY item chance decreased to 8% (from 10%)

fileResolver.py
from resources.colours import *
import random
import pygame

WINDOW_WIDTH = 500
WINDOW_HEIGHT = 500
borderSize = 10


def loadDictionaries(substring):
    indice = [i for i, s in enumerate(dataLocations) if substring in s]

    try:
        with open(dataLocations[indice[0]], 'r') as dictionary:
            result = eval(dictionary.read())
            dictionary.close()
            return result
    except IndexError:
        print('Nothing enumerable, skipping...\nPlease replace empty lines with a #!')


def loadTxt(location):
    with open(location, 'r') as file:
        result = file.read()
        file.close()
        return result


def cRC(limit):  # createRandomCoordinates
    return random.randint(limit / 10, limit - 4 * borderSize)


dataLocations = []
with open('resources/list.txt', 'r') as locationList:
    for line in locationList:
        if '#' not in line:
            dataLocations.append('resources/' + line.strip('\n'))
locationList.close()


mazeRandomiser.py
import random
import numpy


def generator(lines):
    for line in lines:
        for i, cell in enumerate(line):
            requirement = random.randint(0, 100)
            result = random.randint(0, 50)

            if result > requirement:
                line[i] = 1
            if result > 40:
                line[i] = 2

    lines[7][0] = 1
    lines[7][1] = 1
    lines[8][0] = 1
    lines[8][1] = 1
    lines[0][8] = 3

    return lines


def patcher(lines):
    for line in lines:
        for i, cell in enumerate(line):
            try:
                if line[i] == 0:
                    if (line[i - 1] + line[i + 1] + lines[lines.index(line) - 1][i] + lines[lines.index(line) + 1][i]) > 2:
                        if random.randint(1, 3) == 3:
                            line[i] = 1
            except IndexError:
                pass

    return lines


def invert(lines):
    invertedLines = []
    for line in lines:
        newLine = []
        for cell in line:
            if cell == 0:
                cell = 1
            else:
                cell = 0
            newLine.append(cell)
        invertedLines.append(newLine)

    return invertedLines


def solver(invertedLines):
    x = 0
    y = 0

    for i in range(5000):
        prevX = x
        prevY = y
        move = random.randint(0, 3)

        if move == 0 and x != 0:
            x -= 1
        if move == 1 and x != 8:
            x += 1
        if move == 2 and y != 0:
            y -= 1
        if move == 3 and y != 8:
            y += 1

        if x == 8 and y == 8:
            return True
        elif invertedLines[y][x] == 1:
            x = prevX
            y = prevY


def mazeMaker(state):
    while True:
        lines = [1, 2, 3, 4, 5, 6, 7, 8, 9]
        for i, cell in enumerate(lines):
            lines[i] = [0, 0, 0, 0, 0, 0, 0, 0, 0]

        lines = generator(lines)
        lines = patcher(lines)
        lines = numpy.flip(lines, 0)

        invertedLines = invert(lines)
        passes = solver(invertedLines)

        if passes:
            lines = numpy.asarray(lines)
            if state == 'print':
                print('\n')
                for line in lines:
                    print(line)
            return lines


main_preGenerator.py
import pygame
import random
import time
from dictResolver import *
from pygame.locals import *
pygame.init()

COLOURS = loadDictionaries('colours')


class player():
    def __init__(self):
        self.modelWidth = 15
        self.modelHeight = 15
        self.model = pygame.Rect(25, 40, self.modelWidth, self.modelHeight)
        self.colour = COLOURS['WHITE']
        self.vel = 5

        self.projectiles = []
        self.projVel = 8
        self.projColour = COLOURS['WHITE']
        self.projFireRate = 100
        self.lastFireTime = 0

        self.health = 100
        self.healthMax = 100
        self.healthBarBase = pygame.Rect(0, WINDOW_HEIGHT, WINDOW_WIDTH, 50)
        self.healthBar = pygame.Rect(
            0, WINDOW_HEIGHT, self.health * (WINDOW_WIDTH / self.healthMax), 50)
        self.invulnerabilityTime = 0
        self.flashTime = 0

    def updateHealth(self):
        self.healthBar = pygame.Rect(
            0, WINDOW_HEIGHT, self.health * (WINDOW_WIDTH / self.healthMax), 50)

    def checkHealth(self):
        if self.health > self.healthMax:
            self.health = self.healthMax

        self.updateHealth()

        if self.health <= 0:
            for i in range(52):
                x = 0 + i * 5
                WINDOW.fill((x, x, x))
                pygame.display.update()
                time.sleep(0.0512)
            gameOverText = LARGE_FONT.render(
                'Game Over', True, COLOURS['BLACK'], COLOURS['WHITE'])
            textBox = gameOverText.get_rect()
            textBox.center = (WINDOW_WIDTH / 2, WINDOW_HEIGHT / 2)
            WINDOW.blit(gameOverText, textBox)
            pygame.display.update()
            time.sleep(3)
            pygame.quit()

    def fire(self, direction, magnitude):
        self.lastFireTime = TIME
        if direction == 'x':
            xModifier = magnitude * 0.5 * self.modelWidth
            yModifier = 0
        if direction == 'y':
            yModifier = magnitude * 0.5 * self.modelHeight
            xModifier = 0

        self.newProjectile = pygame.Rect(
            self.model.centerx + xModifier, self.model.centery + yModifier, 3, 3)
        self.projectiles.append([self.newProjectile, direction, magnitude])

    def updateProjectiles(self):
        for projectile in self.projectiles:
            if projectile[1] == 'x':
                projectile[0].centerx += projectile[2] * self.projVel
            if projectile[1] == 'y':
                projectile[0].centery += projectile[2] * self.projVel

            if projectile[0].centerx > WINDOW_WIDTH or projectile[0].centerx < 0:
                self.projectiles.remove(projectile)
            if projectile[0].centery > WINDOW_HEIGHT or projectile[0].centery < 0:
                self.projectiles.remove(projectile)


class enemies():
    def __init__(self):
        self.modelWidth = 14
        self.modelHeight = 20
        self.colour = COLOURS['RED']
        self.vel = 4

        self.projectiles = []
        self.projVel = 6
        self.projColour = COLOURS['RED']
        self.projFireRate = 100
        self.lastFireTime = 0

        self.enemyStages = ['DEAD', COLOURS['RED'],
                            COLOURS['AQUA'], COLOURS['MAGENTA']]
        self.screenCleared = {}
        self.currentEnemies = []
        self.currentBosses = []
        self.lastMovement = 0
        self.enemyUpdatePeriod = 50

    def createEnemies(self):
        self.vel = 4 + 0.5 * len(self.screenCleared)

        self.currentEnemies = []
        try:
            for count in range(0, self.enemyCounts[s.currentScreen]):
                self.enemy = pygame.Rect(s.createRandomCoordinates(WINDOW_WIDTH), s.createRandomCoordinates(
                    WINDOW_HEIGHT), self.modelWidth, self.modelHeight)
                if count % 3 == 0:
                    self.type = 'shooter'
                else:
                    self.type = 'base'
                self.currentEnemies.append(
                    [self.enemy, self.colour, self.type, self.lastFireTime])
        except KeyError:
            print('>>> No enemies available for screen {}! @ {}'.format(
                s.currentScreen, TIME))

        if s.currentScreen in self.bosses and self.bosses[s.currentScreen]['health'] > 0:
            self.previousVel = self.vel
            self.bossData = self.bosses[s.currentScreen]
            self.vel = self.bossData['velocity']
            self.type = self.bossData['type']
            self.boss = pygame.Rect(
                WINDOW_WIDTH / 2, WINDOW_HEIGHT / 2, self.bossData['size'], self.bossData['size'])
            self.currentEnemies.append(
                [self.boss, COLOURS['RED'], self.type, self.lastFireTime])
            self.currentBosses.append(
                [self.boss, COLOURS['RED'], self.type, self.lastFireTime])

    def updateEnemies(self):
        for enemyData in self.currentEnemies:
            enemy = enemyData[0]
            for projectile in p.projectiles:
                if projectile[0].colliderect(enemy):
                    try:
                        if enemyData in self.currentBosses:
                            self.bossData['health'] -= 1
                            if self.bossData['health'] <= 0:
                                self.vel = self.previousVel
                                p.health += 20
                                p.updateHealth()
                                self.currentBosses.remove(enemyData)
                                self.currentEnemies.remove(enemyData)
                        else:
                            self.updateHealth(enemyData)
                    except ValueError:
                        print(
                            '>>> Too many projectiles causing too many enemy kills! @ {}'.format(TIME))
                    p.projectiles.remove(projectile)

            if TIME - self.lastMovement > self.enemyUpdatePeriod:
                if enemy.centerx < p.model.centerx:
                    enemy.centerx += self.vel
                else:
                    enemy.centerx -= self.vel
                if enemy.centery < p.model.centery:
                    enemy.centery += self.vel
                else:
                    enemy.centery -= self.vel

        if TIME - self.lastMovement > self.enemyUpdatePeriod:
            self.lastMovement = TIME

        if self.currentEnemies == []:
            self.enemyCounts[s.currentScreen] = 0
            self.screenCleared[s.currentScreen] = True

    def updateHealth(self, enemyData):
        enemyData[1] = self.enemyStages[self.enemyStages.index(
            enemyData[1]) - 1]
        if enemyData[1] == 'DEAD':
            self.currentEnemies.remove(enemyData)

    def checkEnemyCollisions(self):
        p.flashTime = TIME - p.invulnerabilityTime
        for enemyData in self.currentEnemies:
            enemy = enemyData[0]
            if TIME - p.invulnerabilityTime > 500:
                if enemy.colliderect(p.model):
                    if enemyData in self.currentBosses:
                        p.health -= self.bossData['damage']
                        p.invulnerabilityTime = TIME + 500
                    else:
                        p.health -= 20
                        p.invulnerabilityTime = TIME
                p.updateHealth()

        if TIME - p.invulnerabilityTime < 500:
            if round(p.flashTime / 50) % 2 == 0:
                p.colour = COLOURS['BLACK']
            else:
                p.colour = COLOURS['WHITE']
        else:
            p.colour = COLOURS['WHITE']

    def fire(self, enemyData):
        enemy = enemyData[0]
        enemyData[3] = TIME
        if s.currentScreen in e.bosses:
            for bossData in e.currentBosses:
                if enemy == bossData[0]:
                    bossData[3] = enemyData[3]
        dx = p.model.centerx - enemy.centerx
        dy = p.model.centery - enemy.centery

        self.newProjectile = pygame.Rect(enemy.centerx, enemy.centery, 3, 3)
        self.projectiles.append([self.newProjectile, dx, dy])

    def updateProjectiles(self):
        for projectile in self.projectiles:
            self.dirvect = pygame.math.Vector2(projectile[1], projectile[2])
            self.dirvect.normalize()
            self.dirvect.scale_to_length(self.projVel)
            projectile[0].move_ip(self.dirvect)

            try:
                if projectile[0].centerx > WINDOW_WIDTH or projectile[0].centerx < 0:
                    self.projectiles.remove(projectile)
                if projectile[0].centery > WINDOW_HEIGHT or projectile[0].centery < 0:
                    self.projectiles.remove(projectile)

                if projectile[0].colliderect(p.model):
                    p.health -= 10
                    self.projectiles.remove(projectile)
            except ValueError:
                print('Error removing enemy projectile! @ {}'.format(TIME))


class items():
    def checkCurrentScreenItems(self):
        self.currentItems = []
        try:
            for item in self.items[0]:
                if item['chance'] != 0:
                    self.currentItems.append(item)
                    item['rolled'] = 200
        except KeyError:
            print('>>> No items found with all screen spawns! @ {}'.format(TIME))
            pass
        try:
            for item in self.items[s.currentScreen]:
                if item['chance'] != 0:
                    self.currentItems.append(item)
                    item['rolled'] = 200
        except KeyError:
            print('>>> No items found for screen {}! @ {}'.format(
                s.currentScreen, TIME))

    def checkItemCollisions(self):
        for item in self.currentItems:
            if p.model.colliderect(item['model']):
                if 'PLAYER_VELOCITY' in item['attributes']:  # Default 5
                    p.vel += item['magnitude']
                if 'PLAYER_HEALTH' in item['attributes']:  # Default 100
                    p.healthMax += item['magnitude']
                    p.health += item['magnitude']
                if 'PROJECTILE_VELOCITY' in item['attributes']:  # Default 8
                    p.projVel += item['magnitude'] * 2
                if 'PROJECTILE_RATE' in item['attributes']:  # Default 100
                    p.projFireRate -= item['magnitude'] * 5
                print('Player picked up {} @ {}'.format(item, TIME))
                self.currentItems.remove(item)
                item['chance'] = 0


class screens():
    def __init__(self):
        self.currentScreen = 11
        self.update = True

        self.borderColour = (51, 51, 51)
        self.borderSize = 10
        self.topBar = pygame.Rect(0, 0, WINDOW_WIDTH, self.borderSize)
        self.bottomBar = pygame.Rect(
            0, WINDOW_HEIGHT - self.borderSize, WINDOW_WIDTH, self.borderSize)
        self.rightBar = pygame.Rect(
            WINDOW_WIDTH - self.borderSize, 0, self.borderSize, WINDOW_HEIGHT)
        self.leftBar = pygame.Rect(0, 0, self.borderSize, WINDOW_HEIGHT)
        self.lettersToBorders = {'T': self.topBar,
                                 'B': self.bottomBar,
                                 'L': self.leftBar,
                                 'R': self.rightBar
                                 }

    def screenUpdate(self):
        if s.currentScreen % 10 > 3:
            e.colour = COLOURS['AQUA']
        else:
            e.colour = COLOURS['RED']

        i.checkCurrentScreenItems()
        e.createEnemies()
        d.printInfo()

    def drawBorders(self):
        self.bordersToDraw = []
        if e.currentEnemies == []:
            if self.currentScreen in self.borderOverrides and self.bordersToDraw == []:
                for letter in self.borderOverrides[self.currentScreen]:
                    try:
                        self.bordersToDraw.append(
                            self.lettersToBorders[letter])
                    except KeyError:
                        print(
                            '>>> Incorrect borders found for screen border override! @ {}'.format(TIME))
                        pass
            else:
                if self.currentScreen + 10 in e.enemyCounts:
                    self.bordersToDraw.append(self.rightBar)
                if self.currentScreen - 10 in e.enemyCounts:
                    self.bordersToDraw.append(self.leftBar)
                if self.currentScreen + 1 in e.enemyCounts:
                    self.bordersToDraw.append(self.topBar)
                if self.currentScreen - 1 in e.enemyCounts:
                    self.bordersToDraw.append(self.bottomBar)
            for bar in self.bordersToDraw:
                pygame.draw.rect(WINDOW, self.borderColour, bar)
                self.checkBorderCollisions(bar)

    def checkBorderCollisions(self, bar):
        if p.model.colliderect(bar):
            self.screenMoveDirection(bar)
            p.projectiles = []

    def screenMoveDirection(self, bar):
        if bar == self.topBar:
            self.currentScreen += 1
            p.model.centery += WINDOW_HEIGHT - 4 * p.model.height
        if bar == self.bottomBar:
            self.currentScreen -= 1
            p.model.centery -= WINDOW_HEIGHT - 4 * p.model.height
        if bar == self.rightBar:
            self.currentScreen += 10
            p.model.centerx -= WINDOW_WIDTH - 4 * p.model.width
        if bar == self.leftBar:
            self.currentScreen -= 10
            p.model.centerx += WINDOW_WIDTH - 4 * p.model.width

        i.currentItems = []
        self.update = True

    def createRandomCoordinates(self, limit):
        return random.randint(limit / 10, limit - 4 * s.borderSize)


class debug():
    def printInfo(self):
        print('====================' +
              '\nCurrent Screen: {}'.format(s.currentScreen) +
              '\nEntering at X: {}, Y: {}'.format(p.model.centerx, p.model.centery) +
              '\n\nEnemies: {}, Velocity: {}'.format(e.enemyCounts[s.currentScreen], e.vel) +
              '\nBoss: {}, Attributes: {}'.format(self.checkBosses(), self.getBossAttributes()) +
              '\n\nItems: {}'.format(self.getItemAttributes()) +
              '\n\nPlayer Health: {}, Velocity: {}\n\n'.format(p.health, p.vel))

    def checkBosses(self):
        if s.currentScreen in e.bosses:
            return True
        else:
            return False

    def getBossAttributes(self):
        if s.currentScreen in e.bosses:
            return e.bosses[s.currentScreen]
        else:
            return 'None'

    def getItemAttributes(self):
        self.returnString = ''
        for item in i.currentItems:
            if item['chance'] != 0 or item['screen'] == s.currentScreen:
                self.returnString += '\n  {}'.format(item)
        return self.returnString


TIME = 0
FPS = 60
FPSCLOCK = pygame.time.Clock()

WINDOW_WIDTH = 500
WINDOW_HEIGHT = 500
WINDOW = pygame.display.set_mode((WINDOW_WIDTH, WINDOW_HEIGHT + 40))

LARGE_FONT = pygame.font.SysFont('consolas', 64)
NORMAL_FONT = pygame.font.SysFont('consolas', 32)
SMALL_FONT = pygame.font.SysFont('consolas', 16)

p = player()
e = enemies()
i = items()
s = screens()
d = debug()

# Load dictionaries from the 'resources' folder.
e.bosses = loadDictionaries('bosses')
e.enemyCounts = loadDictionaries('enemyCounts')
i.items = loadDictionaries('items')
s.borderOverrides = loadDictionaries('borderOverrides')

while True:
    pygame.time.delay(round(1000 / FPS))

    if s.update == True:
        s.screenUpdate()
        s.update = False

    keys = pygame.key.get_pressed()

    if keys[pygame.K_LEFT] and p.model.left > p.vel:
        p.model.left -= p.vel
    if keys[pygame.K_RIGHT] and p.model.right < WINDOW_WIDTH - p.vel:
        p.model.right += p.vel
    if keys[pygame.K_UP] and p.model.top > p.vel:
        p.model.top -= p.vel
    if keys[pygame.K_DOWN] and p.model.bottom < WINDOW_HEIGHT - p.vel:
        p.model.bottom += p.vel

    if TIME - p.lastFireTime >= p.projFireRate:
        if keys[pygame.K_w]:
            p.fire('y', -1)
        elif keys[pygame.K_s]:
            p.fire('y', 1)
        elif keys[pygame.K_a]:
            p.fire('x', -1)
        elif keys[pygame.K_d]:
            p.fire('x', 1)

    p.updateProjectiles()
    e.updateEnemies()
    e.updateProjectiles()
    e.checkEnemyCollisions()
    i.checkItemCollisions()

    WINDOW.fill(COLOURS['BLACK'])
    for enemyData in e.currentEnemies:
        pygame.draw.rect(WINDOW, enemyData[1], enemyData[0])
        if enemyData[2] == 'shooter' and TIME - enemyData[3] > 500:
            e.fire(enemyData)

    for item in i.currentItems:
        if item['rolled'] == 200:
            item['rolled'] = random.randint(0, 100)
        if e.currentEnemies == []:
            if (item['chance'] >= item['rolled'] and item['screen'] == 0) or item['screen'] == s.currentScreen:
                item['screen'] = s.currentScreen
                pygame.draw.rect(
                    WINDOW, COLOURS[item['colour']], item['model'])

    for projectile in p.projectiles:
        pygame.draw.rect(WINDOW, p.projColour, projectile[0])
    for projectile in e.projectiles:
        pygame.draw.rect(WINDOW, e.projColour, projectile[0])

    s.drawBorders()
    pygame.draw.rect(WINDOW, p.colour, p.model)
    pygame.draw.rect(WINDOW, COLOURS['RED'], p.healthBarBase)
    pygame.draw.rect(WINDOW, (57, 255, 14), p.healthBar)
    p.checkHealth()
    pygame.display.update()

    FPSCLOCK.tick(FPS)
    TIME += FPSCLOCK.tick(FPS)

    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
